### DISCUSSION TITLE
Skip ahead, or run part of, a Gcode file/job

### DISCUSSION THREAD

**Discussion by joakimgk** [winder/Universal-G-Code-Sender#2593] [Ideas]
Would it be possible to have a feature where one could run only part of a gcode "job"? I'm using the software to make PCBs, where the gcode file is a set of traces defined by FlatCam. Often, the job is less successful in regions of the PCB board than others, so it would be very convenient to be able to either

a) draw a region (rectangle?) on the Visualizer, and say "run only this part", or
b) be able to skip ahead to next trace while working.

The FlatCam top/bottom layer jobs (gcode files) consists of disjoint traces. Perhaps it might be possible to somehow skip on to the next such trace?

I hope I'm explaining this properly. If it makes sense, or is feasible, is another question :D 

Maybe a third suggestion could be 

c) be able to adjust XYZ zero while the job is paused... So one might be able to nudge the job a bit higher/deeper (Z-axis) etc in certain regions.

I am aware of the heightmap (AutoLeveler) feature of the software, and I am using it (with a probe). However, I still experience inconsistent carving across the PCB. Disclaimer: 100% surely this is my fault, and me not being thorough enough in preparations etc, but I still wanted to pitch those ideas. Great software!

**Comment by breiler** [comment]
Hi Joakim,

I think that this is a clear and descriptive use case. I does however sound difficult to solve. Below is my quick thoughts on this, if anyone else has any inputs feel free to add!

a) This could be possible. Today we have the RunFromProcessor which takes the line number where it should start. Its job is to figure out the starting point of the machine by parsing through each skipped gcode line and accumulate the start position. It should be possible to extend this with a list of from-to line numbers. Do note however, that for each intersection of the path with the selection-rectangle, UGS would need to add a rapid move to a safe height and then a rapid move to the next start intersection point. There will be a lot of extra rapid movements which could make this unusable.

b) there is no concept of a "trace" in the gcode so I can't see how this would be possible.

c) I would say no. As I see it there are be two possible ways to solve this. One would be to lower the work offset position with either G92 or G10 command during runtime. But since the controller is buffering a couple of commands in planner buffer there would be a bit of lag for these changes to take effect. And if the gcode program or the current setup is utilizing temporary work offsets, this would overwrite those and probably crash the machine. Another way would be to intercept any Z move command and subtract the wished amount. But then there is the problem with absolute (G90) and relative (G91) positioning. It will be very difficult to get this right and cover all possible scenario.

**Reply by joakimgk** [reply]
Thanks for the constructive response :) 
When I say "trace" in point "b", I mean the software would have to compute all connected lines comprising separate objects. Such an object could be any shape where the machine runs without lifting Z to safe/transport height. I'm just refering to this as "trace" in the context of FlatCam/PCB making, since these jobs do consist entirely of such separate closed objects. But, of course, for a feature to make sense it would have to be sensible for any kind of CNC job, and maybe PCB carving (traces) is a very special kind of job. Guess it's not really what these machines are made for ;)
